### PR TITLE
Rewrite faucet with tokio v0.3

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Faucet code still depends on tokio v0.1

#### Summary of Changes
Rewrite using tokio v0.3 to synchronize with the majority of the code base
~~Upgrading to tokio v1.0 will require bumping `bytes` and `tokio-util` as well~~ 🎉 

Toward #10187

### CODE REVIEW

File: faucet/src/faucet.rs [hunk 1]
```diff
@@ -283,52 +284,63 @@ pub fn run_local_faucet(
...
+    }
+}
 
-            let processor = reader.and_then(move |bytes| {
-                match faucet2.lock().unwrap().process_faucet_request(&bytes) {
+async fn process(
+    stream: TokioTcpStream,
+    faucet: Arc<Mutex<Faucet>>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let framed = BytesCodec::new().framed(stream);
```
Comment (t-nelson) [inline faucet/src/faucet.rs:321]: Unrelated to these changes, but I wonder if we have an unbounded read here, like we did in ip-echo-server (#13568)

Comment (CriesofCarrots) [reply faucet/src/faucet.rs:321]: Looks like. Thanks for pointing out that PR!
Shouldn't be an issue to do an read_exact here. I'll plug that in shortly.

Comment (CriesofCarrots) [reply faucet/src/faucet.rs]: @t-nelson can you please take a peek at https://github.com/solana-labs/solana/pull/14336/commits/fbd0645ff179647c63acb6e4788322ab8177d1ae?

**Review by t-nelson — APPROVED** [review]
Hey!  Where'd all those dependencies go!? :sunglasses: 

LGTM!

**Review by mvines — APPROVED** [review]

**Pull request merged**